### PR TITLE
Add Google OAuth2 authentication flow 

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

 
Adds an OAuth2 authentication flow, as a preparation for the upcoming GAM integration.
Currently, Site Kit's authorisation is reused to access user's GA data. Site Kit does not support the scopes necessary to access GAM data, so Newspack has to roll its own authorisation for this purpose.  

UI: I have made it very lightweight purposefully, since this PR is quite large as it is. The UI should probably live in some kind of settings view. cc @thomasguillot 

### How to test the changes in this Pull Request:

1. Nothing should change unless the necessary environment variables are present, so have a look at the Dashboard, Campaigns Analytics, and Analytics wizard - everything should work as before
2. Create a [Google OAuth2 app](https://support.google.com/cloud/answer/6158849?hl=en). Set `<test-site->/wp-admin/admin.php?page=newspack` (e.g. `https://newspack.test/wp-admin/admin.php?page=newspack`) as an Authorized redirect URI.
3. Set client ID and secret of the app as `NEWSPACK_GOOGLE_OAUTH_CLIENT_ID` and `NEWSPACK_GOOGLE_OAUTH_CLIENT_SECRET` environment variables in `wp-config.php`
4. Visit the Dashboard, observe a new section above the wizard links - with an "Authorise" button
5. Click the button and complete the authorisation flow - observe a success message and information "Authorized Google as <email>" in the end
6. This was the happy path, now let's test how it can break. Remove the `_newspack_google_oauth` user meta from the DB
7. Reload the Dasboard, observe initial (pre-authorisation) state. Complete the authorisation from again.
8. Visit your [Google account's permissions screen](https://myaccount.google.com/permissions) and revoke permissions for the app
9. Again, observe the pre-authorisation state and complete the flow again.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by thomasguillot** [comment]
@adekbadek I moved the OAuth2 section to the bottom of the dashboard.

Screenshot 2021-04-16 at 16 31 44

Not quite sure but `goToAuthPage` didn't seem to work... not quite sure why 🤔

> UI: I have made it very lightweight purposefully, since this PR is quite large as it is. The UI should probably live in some kind of settings view

Do you mean the actual connection, or the client ID/secret?

**Comment by adekbadek** [comment]
> > UI: I have made it very lightweight purposefully, since this PR is quite large as it is. The UI should probably live in some kind of settings view

> Do you mean the actual connection, or the client ID/secret?

The connection - it's better after your changes, but still it's an odd item among links to wizards.

**Comment by adekbadek** [comment]
> Not quite sure but goToAuthPage didn't seem to work... not quite sure why 🤔

What happens after you click on the authorisation dashboard item?

**Comment by thomasguillot** [comment]
> What happens after you click on the authorisation dashboard item?

Absolutely nothing

**Comment by thomasguillot** [comment]
> The connection - it's better after your changes, but still it's an odd item among links to wizards.

Oh ok I see.
What if we replicate something the integrations step from the onboarding? Where a user could see Jetpack, Site Kit, Mailchimp, and Google OAuth statuses

**Comment by adekbadek** [comment]
> What if we replicate something like the integrations step from the onboarding?

Definitely, but that's on top of the dashboard/post-onboarding UI.

File: includes/google/class-google-services-connection.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	 *
+	 * @return Authentication Site Kit's authentication.
+	 */
+	public static function get_site_kit_authentication() {
+		if ( defined( 'GOOGLESITEKIT_PLUGIN_MAIN_FILE' ) ) {
+			$context        = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+			$authentication = new Authentication( $context, new Options( $context ), new User_Options( $context ) );
+			return $authentication;
+		} else {
+			return new WP_Error( 'newspack_sitekit_disconnectedt', __( 'Connect Site Kit plugin to view Campaign Analytics.', 'newspack' ) );
```
Comment (dkoo) [inline includes/google/class-google-services-connection.php:36]: Typo?

```suggestion
			return new WP_Error( 'newspack_sitekit_disconnected', __( 'Connect Site Kit plugin to view Campaign Analytics.', 'newspack' ) );
```

**Review by dkoo — APPROVED** [review]
Really like the consolidation of the Google Services authorizations into a single class, and the DRY-ability(?) it enables. I have a few nits below but nothing blocking, so 👍

File: README.md [hunk 2]
```diff
@@ -30,12 +30,17 @@ If you have a patch or have stumbled upon an issue with the Newspack plugin/them
...
-define('NEWSPACK_WPCOM_CLIENT_ID', '12345');
-// optional
-define('NEWSPACK_SUPPORT_IS_PRE_LAUNCH', true);
+// Support
+define( 'NEWSPACK_SUPPORT_API_URL', 'https://super-tech-support.zendesk.com/api/v2' );
+define( 'NEWSPACK_SUPPORT_EMAIL', 'support@company.com' );
+define( 'NEWSPACK_WPCOM_CLIENT_ID', '12345' );
+
+// OAuth2
+define( 'NEWSPACK_GOOGLE_OAUTH_CLIENT_ID', 'abc' );
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply README.md:39]: It might take a while, and also I don't think it's pressing to expose that part of the configuration to regular users.

**Review by thomasguillot — APPROVED** [review]

**Pull request merged**